### PR TITLE
Implement challenge problem: maximize optional attendees within constraints of mandatory attendee availability.

### PR DESCRIPTION
**Pull request by dtjanaka**
_Implement an optimized version of the optional attendee functionality: If no time exists for all optional and mandatory attendees, find the time slot(s) that allow mandatory attendees and the greatest possible number of optional attendees to attend._

- Implemented above functionality to find slots for mandatory attendees and maximum number of optional attendees.
  - After finding the slots for mandatory attendees, find all contained durations as long as the requested meeting length and count the number of optional attendees who can attend each
  - Return those durations which allow the maximum number of optional attendees 
- Added tests for maximum optional attendees scenarios and duration edge cases (zero, negative).

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -18,18 +18,138 @@
...
+  /**
+   * Return the TimeRanges available for all meeting members given existing
+   * events
+   */
+  public Collection<TimeRange> query(Collection<Event> events,
+                                     MeetingRequest request) {
+    // For a meeting longer than a day, no times are possible
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
```
Comment (robotsaurus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: What about negative duration?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -18,18 +18,138 @@
...
+    // meeting of the desired duration, store the number of optional attendees 
+    // who can attend
+    ArrayList<ArrayList<Integer>> forEachDuration =
+        new ArrayList<ArrayList<Integer>>();
+    
+    for (int i = 0; i < mandatoryAvailability.size(); i++) {
+      forEachDuration.add(new ArrayList<Integer>());
+      int n = 0;
+      while (n * increment + duration <=
+             mandatoryAvailability.get(i).duration()) {
```
Comment (robotsaurus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: I'm not following why you are increasing the duration by n*increment?

Don't you want to check if the meeting starts at 0:1440 and lasts the duration how many people can go?

**Review by robotsaurus — CHANGES_REQUESTED** [review]

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (dtjanaka) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This shifts the duration by `increment` minutes until the end of the slot in the mandatory attendees' calendar.
```
|--|    increment
|-----| duration

|-----|          Duration-length slot to check availability of all optional attendees, n = 0
|-------------|  Time slot where all mandatory attendees are available

|--|-----|       Duration-length slot to check availability of all optional attendees, n = 1
|-------------|  Time slot where all mandatory attendees are available
.
.
.
|--|--|--|-----| n++ until no more duration-length slots possible
|-------------|  Time slot where all mandatory attendees are available
```
If I understand what you're saying about 0:1440, I have not accounted for midnight represented as either 0 or 1440? The TimeRange class defines a day as going from 0 to 1439 minutes, and my code works for a meeting potentially starting at 0 minutes.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -18,18 +18,138 @@
...
+
+        int numCanGo = 0;
+        for (ArrayList<TimeRange> curPersonCal : optionalAttendeesCals) {
+          for (TimeRange curTime : curPersonCal) {
+            int incrementedStart = 
+                mandatoryAvailability.get(i).start() + n * increment;
```
Comment (robotsaurus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: This gets recomputed multiple times for the double for loop but never changes.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (robotsaurus) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Oh, I wasn't asking about midnight that was just shorthand for the range [0,1,2,3,....,1440].  

Ah, now I understand, I thought you were changing the duration.

I think this would be more clear if you changed this to a for loop and pulled out the incrementedStart that I commented out below.  So something like:

`
// Iterate over all the potential meeting blocks of size duration and count the number of optional attendees that are free.

for (int potentialStartTime = mandatoryAvailability.get(i).start(); potentialStartTime < mandatoryAvailability.get(i).end() - duration; potentialStartTime++) {
   ...
}
`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -18,18 +18,138 @@
...
+    // E.g. if all meetings are only held at 15 minute increments of the hour
+    // for durations that are multiples of 15 minutes, can reduce amount of
+    // computation by setting {increment} to 15
+    int increment = 1;
+
+    // For all possible times every {increment} minutes that allows for a 
+    // meeting of the desired duration, store the number of optional attendees 
+    // who can attend
+    ArrayList<ArrayList<Integer>> forEachDuration =
+        new ArrayList<ArrayList<Integer>>();
```
Comment (robotsaurus) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Using the name forEachDuration as well as "meeting of the desired duration" also caused confusion.  This isn't the meeting duration but rather the potential start time.

**Review by robotsaurus — CHANGES_REQUESTED** [review]

**Comment by dtjanaka** [comment]
PTAL

**Review by robotsaurus — APPROVED** [review]
